src/partition.mjs

```diff
@@ -1,30 +1,39 @@
 import Dequeue from 'dequeue'
-import iterable from './iterable'
+import ensureIterable from './internal/ensure-iterable'
 
-function uncurried (func, iter) {
+function partition (func, iter) {
   const satisfied = new Dequeue()
   const unsatisfied = new Dequeue()
-  const iterator = iterable(iter)[Symbol.iterator]()
+  const iterator = ensureIterable(iter)[Symbol.iterator]()
+  let exhausted = 0
 
   function * part (queue) {
-    while (true) {
-      while (queue.length) {
-        yield queue.shift()
-      }
+    try {
+      while (true) {
+        while (queue.length) {
+          yield queue.shift()
+        }
 
-      const { value, done } = iterator.next()
-      if (done) break
+        const { value, done } = iterator.next()
+        if (done) break
 
-      const chosen = func(value) ? satisfied : unsatisfied
-      chosen.push(value)
+        const chosen = func(value) ? satisfied : unsatisfied
+        chosen.push(value)
+      }
+    } finally {
+      exhausted++
+      if (exhausted === 2) {
+        if (typeof iterator.return === 'function') iterator.return()
+      }
     }
   }
 
   return [part(satisfied), part(unsatisfied)]
 }
 
-const curried = func => iter => uncurried(func, iter)
-
-export default function partition (func, iter) {
-  return iter === undefined ? curried(func) : uncurried(func, iter)
+export default function curriedPartition (func, iter) {
+  if (typeof iter === 'undefined') {
```

Comment:
What's the difference between `x === undefined` and `typeof x === 'undefined'`? I think `x === undefined` should be faster since it does not deal with string comparison 🤔.

Reply:
It is an old js baggage. Due to the fact that you can sadly redefine undefined.
But I can see now, that is not recommended anymore http://2ality.com/2013/04/check-undefined.html
I will reverse that

Reply:
> I will reverse that

No needs if you want to stay consistent.

But if you really want to use `undefined`, I think you should also make similar changes in other places.

Reply:
I'll review the code base afterwards